### PR TITLE
[amp-refactor][7/n] Refactor AssetDaemonAssetCursor -> AssetConditionCursor

### PR DESCRIPTION
**Pull request by OwenKephart**
## Summary & Motivation

This PR renames the AssetDaemonAssetCursor object to the more generic `AssetConditionCursor` (imagine a future world in which AssetConditions are possible to evaluate outside the context of the AssetDaemon). It also removes the highly-specific `materialized_requested_and_discarded_subset` and replaces it with the general purpose "extras" field, which allows each condition to optionally stash arbitrary serializable information in a dictionary.

This doesn't yet make changes to the serialized format, so we have to generate these things based off of what we do actually have stored in the cursor, namely the latest AssetConditionEvaluation object for the asset and the 'handled subset', which is used by the MaterializeOnMissing rule (hence the weird hard-coding stuff, which is removed in the next PR in this stack)

### CODE REVIEW

**Comment by OwenKephart** [comment]
Current dependencies on/for this PR:
* `master`
  * **PR #18612** <a href="https://app.graphite.dev/github/pr/dagster-io/dagster/18612?utm_source=stack-comment-icon" target="_blank">Graphite</a> 
    * **PR #18613** <a href="https://app.graphite.dev/github/pr/dagster-io/dagster/18613?utm_source=stack-comment-icon" target="_blank">Graphite</a> 
      * **PR #18614** <a href="https://app.graphite.dev/github/pr/dagster-io/dagster/18614?utm_source=stack-comment-icon" target="_blank">Graphite</a> 
        * **PR #18615** <a href="https://app.graphite.dev/github/pr/dagster-io/dagster/18615?utm_source=stack-comment-icon" target="_blank">Graphite</a> 
          * **PR #18656** <a href="https://app.graphite.dev/github/pr/dagster-io/dagster/18656?utm_source=stack-comment-icon" target="_blank">Graphite</a> 
            * **PR #18684** <a href="https://app.graphite.dev/github/pr/dagster-io/dagster/18684?utm_source=stack-comment-icon" target="_blank">Graphite</a> 
              * **PR #18693** <a href="https://app.graphite.dev/github/pr/dagster-io/dagster/18693?utm_source=stack-comment-icon" target="_blank">Graphite</a>  👈

This [stack of pull requests](https://stacking.dev/?utm_source=stack-comment) is managed by [Graphite](https://app.graphite.dev/github/pr/dagster-io/dagster/18693?utm_source=stack-comment).

File: python_modules/dagster/dagster/_core/definitions/asset_daemon_cursor.py [hunk 1]
```diff
@@ -15,26 +17,116 @@
...
+
+_PLACEHOLDER_HANDLED_SUBSET_KEY = "handled_subset"
+
+
+class AssetConditionCursorExtras(NamedTuple):
+    """Class to represent additional unstructured information that may be tracked by a particular
+    asset condition.
+    """
+
+    condition_snapshot: "AssetConditionSnapshot"
```
Comment (sryza) [inline python_modules/dagster/dagster/_core/definitions/asset_daemon_cursor.py:53]: Why put this on this object instead of on the asset condition cursor?

Comment (OwenKephart) [reply python_modules/dagster/dagster/_core/definitions/asset_daemon_cursor.py:53]: The idea is that the AssetConditionCursor is _not_ a recursive object -- it just has a flat set of properties.

One of those properties is a list of these extras objects, which are mapped to a specific sub-expression. So the condition_snapshot here is the snapshot of one of the child conditions, whereas the condition_snapshot on the cursor is the snapshot of the root condition.

Another option would be to just formalize this mapping structure and do one of those NonScalarSerializableMapping things where we have a mapping from ConditionSnapshot to Mapping[str, PackableValue].

That might be a cleaner solution as it avoids adding an extra whitelisted class

File: python_modules/dagster/dagster/_core/definitions/asset_daemon_cursor.py [hunk 2]
```diff
@@ -15,26 +17,116 @@
...
+    extras: ExtrasDict
```
[the remark this answers is not in the document]
Comment (OwenKephart) [reply python_modules/dagster/dagster/_core/definitions/asset_daemon_cursor.py:54]: Hm I see what you mean about just allowing a PackableValue (is that just a superset of allowing any whitelisted object)?

I don't particularly see any issue with that approach -- ultimately it seems nicer than having hardcoded strings for accessing specific bits of data, so I can change to that direction.

[the remark this answers is not in the document]
Comment (sryza) [reply python_modules/dagster/dagster/_core/definitions/asset_daemon_cursor.py:54]: > (is that just a superset of allowing any whitelisted object)?

Ah oops, I though `PackableValue` is what `JsonSerializableValue` actually is.

File: python_modules/dagster/dagster/_core/definitions/asset_daemon_cursor.py (hunk 1 above)
Comment (sryza) [reply python_modules/dagster/dagster/_core/definitions/asset_daemon_cursor.py:53]: Ahh I see. I think I got tripped up by different interpretations of the plural "extras".

I do think a mapping would make it clearer. Putting `descendants` in the name of the attribute on `AssetConditionCursor` would also make it clearer.

**Comment by OwenKephart** [comment]
@sryza A small rebasing disaster occurred when attempting to address the issue at hand here, resulting this branch and a couple others getting mixed up by graphite.

The end result is that the contents of this PR have been essentially duplicated by https://github.com/dagster-io/dagster/pull/18953, so I'm closing this one. The refactor to change how we process the extras value was then plugged into a PR in which we make other changes to the return type of the evaluate() method: https://github.com/dagster-io/dagster/pull/18823

**Pull request closed without merging**